### PR TITLE
JS: Add ClientRequest.getAResponseDataNode

### PR DESCRIPTION
**Pull request by asger-semmle**
Adds a way to get the response from a client request.

It also exposes how the response is represented using a `responseType` string and a `promise` boolean, mainly motivated by the need to distinguish JSON objects from strings and without relying too much on local data flow. For example I didn't want to rely on a promise being `then`-ed in the same function that sends the request, so it may return a node that refers to a promise of the response.

Handling the different response types ended up being more difficult than I anticipated and the end result mostly works but is somewhat imperfect. For example, `axios` returns a promise of an object whose `data` property is the response stream. The model just points to the returned promise as the response, with no indication that the `data` property but be read to get to the response.

Ultimately I'm not sure if it was worth the hassle, comments welcome.

One other thing is that the different kinds of client requests are now public, but hidden inside the `ClientRequest::` namespace. This makes it easier to extend the models while reusing the bits we already have in place. For example, one might want to extend `getAnEventListener` to use type tracking.

Commit-by-commit review is recommended despite the messy commit history.

### CODE REVIEW

**Comment by esben-semmle** [comment]
> The model just points to the returned promise as the response, with no indication that the data property but be read to get to the response.

Is it only the `resp.data` property that is controlled by the remote server?
You could add another flowstep to the model such that this `data` property is the result for `getAResponseDataNode`.

(If you care for both `resp` and `resp.data` you can let both be the result (with appropriate `responseTypes`))

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 1]
```diff
@@ -152,6 +219,24 @@ private class AxiosUrlRequest extends ClientRequest::Range {
       result = getOptionArgument([0 .. 2], name)
     )
   }
+
+  string getResponseFormat() {
+    exists(DataFlow::Node option | option = getOptionArgument([0 .. 2], "responseType") |
+      result = option.getStringValue()
+      or
+      not exists(option.getStringValue()) and
+      result = ""
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: Can we add a test for the `""` response? It is a general gotcha to accidentally remove the "unknown" result.

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 2]
```diff
@@ -37,6 +37,36 @@ class ClientRequest extends DataFlow::InvokeNode {
...
+    * namely `arraybuffer`, `blob`, `document`, `json`, or `text`.
+    *
+    * Additionally, the `responseType` may have one of the following values:
+    * - `fetch.response`: The result is a `Response` object as defined by the [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Response).
+    * - `stream`: The result is a Node.js stream
+    * - `error`: The result is an error in an unspecified format, possibly containing information from the response
+    *
+    *
+    * Custom implementations of `ClientRequest` may use other formats.
+    * If the responseType is not known the convention is to use an empty string.
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: I take that you already have decided that a `FlowLabel`-style enumeration of all response types using `class ResponseType extends string` is architectural overkill?

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 3]
```diff
@@ -152,6 +219,24 @@ private class AxiosUrlRequest extends ClientRequest::Range {
...
+  string getResponseFormat() {
+    exists(DataFlow::Node option | option = getOptionArgument([0 .. 2], "responseType") |
+      result = option.getStringValue()
+      or
+      not exists(option.getStringValue()) and
+      result = ""
+    )
+    or
+    not exists(getOptionArgument([0 .. 2], "responseType")) and
+    result = "json"
```
[the remark this answers is not in the document]
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: We don't just blindly assume "json" - it's the default value if not specified. The option object could come from a parameter, in which case we'd just assume the default. I've updated this to be more conservative although I feel it became very verbose in the process.

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll (hunk 2 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: Just not seeing the benefit.

**Comment by asger-semmle** [comment]
> Is it only the resp.data property that is controlled by the remote server?

There's also the `statusText` and `headers` properties.

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 4]
```diff
@@ -37,8 +37,41 @@ class ClientRequest extends DataFlow::InvokeNode {
...
   DataFlow::Node getADataNode() { result = self.getADataNode() }
+
+  /**
+   * Gets a data flow node that refers to some representation of the response, possibly
+   * wrapped in a promise object.
+   *
+   * The `responseType` describes how the response is represented as a JavaScript value
+   * (after resolving promises), and may assume the following values:
+   * - Any response type defined by [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseType):
+   *    - `text`: The result is a string
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: Super minor nit: the whitespace before the list items is inconsistent:

```
   * - Any response type defined by [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseType):
   *    - `text`: The result is a string
```
vs
```
   * - Any of the following additional response types defined by this library:
   *  - `fetch.response`: The result is a `Response` object from [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Response).
```

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 5]
```diff
@@ -37,8 +37,41 @@ class ClientRequest extends DataFlow::InvokeNode {
...
+   *    - `json`: The result is a deserialized JSON object
+   *    - `arraybuffer`: The result is an `ArrayBuffer` object
+   *    - `blob`: The result is a `Blob` object
+   *    - `document`: The result is a deserialized HTML or XML document
+   * - Any of the following additional response types defined by this library:
+   *  - `fetch.response`: The result is a `Response` object from [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Response).
+   *  - `stream`: The result is a Node.js stream and `http.IncomingMessage` object
+   *  - `header`: The result the value of a header, as a string
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]:
```suggestion
   *  - `header`: The result is the value of a header, as a string
```

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 6]
```diff
@@ -62,222 +95,465 @@ module ClientRequest {
...
 
-/**
- * Gets the name of a property that likely contains a URL value.
- */
-private string urlPropertyName() {
-  result = "uri" or
-  result = "url"
-}
+  /**
+  * Gets name of an HTTP request method, in all-lowercase.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]:
```suggestion
  * Gets the name of an HTTP request method, in all-lowercase.
```

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 7]
```diff
@@ -62,222 +95,465 @@ module ClientRequest {
...
+    predicate isJson() {
+      getOptionArgument(1, "json").mayHaveBooleanValue(true)
+    }
+
+    override DataFlow::Node getAResponseDataNode(string responseType, boolean promise) {
+      result = this and
       (
-        callee = DataFlow::moduleImport(moduleName) or
-        callee = DataFlow::moduleMember(moduleName, "stream")
+        isStream() and
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: You could rephrase this as `if isStream() then ... else if isJson() ... else ...`, which would also model the fact that the `json` flag is [ignored](https://www.npmjs.com/package/got#json) in streaming mode.

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 8]
```diff
@@ -62,222 +95,465 @@ module ClientRequest {
...
-private class ClosureXhrIoRequest extends ClientRequest::Range {
-  ClosureXhrIoRequest() {
-    exists(DataFlow::SourceNode xhrIo | xhrIo = Closure::moduleImport("goog.net.XhrIo") |
-      this = xhrIo.getAMethodCall("send")
+    /**
+     * Gets a node that refers to this XHR object.
+     *
+     * In particular, this can be the receiver of an event handler.
+     */
+    DataFlow::SourceNode getAnAlias() {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: `private`?

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 9]
```diff
@@ -62,222 +95,465 @@ module ClientRequest {
...
+      )
+    }
+
+    override DataFlow::Node getUrl() { result = getArgument(0) }
+
+    override DataFlow::Node getHost() { none() }
+
+    override DataFlow::Node getADataNode() {
+      result = getArgument(2) or
+      result = getArgument(3)
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: `result = getArgument([2 .. 3])`

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 10]
```diff
@@ -62,222 +95,465 @@ module ClientRequest {
...
+      )
+    }
 
-  override DataFlow::Node getHost() { none() }
+    /**
+     * Gets a node that refers to this XHR object.
+     *
+     * In particular, this can be the receiver of an event handler.
+     */
+    DataFlow::SourceNode getAnAlias() {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: `private`?

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll [hunk 11]
```diff
@@ -62,222 +95,465 @@ module ClientRequest {
...
+      result = base
+      or
+      result = getAnEventListener().getReceiver()
+    }
 
-  override DataFlow::Node getADataNode() {
-    result = getArgument(2) or
-    result = getArgument(3)
+    /**
+     * Gets the response type corresponding `getReponse()` but not
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]:
```suggestion
     * Gets the response type corresponding to `getReponse()` but not
```

(?)

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll (hunk 8 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: I'd like to be able to override it if non-local aliasing is needed.

File: javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll (hunk 10 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll]: ditto
